typing/typedecl_variance.ml

```diff
@@ -300,6 +300,10 @@ let compute_variance_decl env ~check decl (required, _ as rloc) =
       compute_variance_type env ~check rloc decl
         (mn @ List.map (fun {Types.ld_mutable; ld_type} ->
              (ld_mutable = Mutable, ld_type)) ftl)
+  in
+  if mn = [] || decl.type_kind <> Type_abstract then
+    List.map Variance.strengthen vari
+  else vari
```

Comment:
@garrigue: I forgot why we are disabling relaxed generalization when a covariant variable is in non-strictly-positive position (here you are generalizing again covariant variables in non-strictly-positive but injective positions). Is there a reference that explains this, or at least gives the relevant unsoundness example? (Is it written in a .mli or somewhe, and should it?)

Reply:
This is explained in the original paper on the relaxed value restriction, which you can find on my home page.
  https://www.math.nagoya-u.ac.jp/~garrigue/papers/morepoly-long.pdf
The details are at beginning of page 9.

Reply:
I didn't find anything at the beginning of page 9, but page 15 explains:

> The definition of V − captures more variables than the usual
definition of con- travariant occurrences. We deem dangerous all
occurrences appearing in a contravariant branch of a type. While
this is not necessary to ensure type soundness, we need it to
keep principality of type inference. For instance, consider the
following function.
>
>     let f = let r = ref nil in λk.Y (λf.f ) !r
>
> As the type of Y (λf.f) is ∀αβ.α → β, we expect the principal
type of f to be ∀β.γ → β, with γ a non generalizable
variable. However, if we were to generalize covariant variables
at ranks higher than 0, then ∀βδ.(δ → γ) → β would be another
acceptable type for f, and neither of the two is an instance of
the other. i.e. we would have lost principality.

(Note: I realize now the same discussion is at the beginning of page 9 of the non-extended version of the paper: https://caml.inria.fr/pub/papers/garrigue-value_restriction-fiwflp04.pdf , which is probably the source of the confusion. )

The example translates into OCaml as follows:

```ocaml
let f = let r = ref [] in (fun _ _ -> assert false) !r;;
(* val f : '_weak1 -> 'a *)
```

Here is my understanding of the argument: consider any term whose inferred type is of the form `'_weak -> foo`, where `'_weak` is a local inference variable not shared with the context (it would be generalizable if the term was a value). Then we could always instantiate `'_weak` with `'_w1 -> '_w2`, so this term also has the type `('_w1 -> '_w2) -> foo`, where `'w1` is now in covariant (but not strictly positive) position. If we did allow generalization of variables in covariant-but-not-strictly-positive positions, then we could generalize this into the type `forall 'a. ('a -> '_w2) -> foo`; but that type is not an instance of our original `'_weak -> foo` with the usual ML instance relation.

I still think that it could be useful to explain something like this somewhere in the description of our variances. It's a short explanation, and it's less work than trying to find it in the middle of the paper.